ouroboros-consensus/test-storage/Test/Ouroboros/StateMachine.hs

```diff
@@ -0,0 +1,462 @@
+    deriving (Show, Generic)
+
+instance ToExpr (Model Symbolic)
+instance ToExpr (Model Concrete)
+
+-- This boolean is a hackish way to inticate that any action
+-- needs to take place on a new folder.
+data Command (r :: Type -> Type)
+    = Open                        Bool FsPath IOMode
+    | Close                       Bool (Fd, (Reference (Opaque (FsHandle IOFSE)) r))
+    | Seek                        Bool (Fd, (Reference (Opaque (FsHandle IOFSE)) r)) SeekMode Word64
+    | Get                         Bool (Fd, (Reference (Opaque (FsHandle IOFSE)) r)) Int
+    | Put                         Bool (Fd, (Reference (Opaque (FsHandle IOFSE)) r)) BL.ByteString
+--    | Truncate                    Bool (Fd, (Reference (Opaque (FsHandle IOFSE)) r)) Word64
+    | CreateDirectory             Bool FsPath
+    | CreateDirectoryIfMissing    Bool Bool FsPath
+    | ListDirectory               Bool FsPath
+    | DoesDirectoryExist          Bool FsPath
+    | DoesFileExist               Bool FsPath
+      deriving (Generic1, Rank2.Foldable, Rank2.Traversable, Rank2.Functor)
+
+deriving instance Show (Command Symbolic)
+deriving instance Show (Command Concrete)
+
+data Response (r :: Type -> Type)
+  = OpenR                            (Either FsError (Reference (Opaque (FsHandle IOFSE)) r))
+  | CloseR                           (Either FsError ())
+  | SeekR                            (Either FsError Word64)
+  | GetR                             (Either FsError ByteString)
+  | PutR                             (Either FsError Word64)
+--  | TruncateR                        (Either FsError ())
+  | CreateDirectoryR                 (Either FsError ())
+  | CreateDirectoryIfMissingR        (Either FsError ())
+  | ListDirectoryR                   (Either FsError [String])
+  | DoesDirectoryExistR              (Either FsError Bool)
+  | DoesFileExistR                   (Either FsError Bool)
+  deriving (Generic1, Rank2.Foldable)
+
+deriving instance Show (Response Symbolic)
+deriving instance Show (Response Concrete)
+
+initModel :: Model r
+initModel = Model . Opaque $ (newEmptyMockFS, 0, 0, Map.empty)
+
+preconditions :: Model Symbolic -> Command Symbolic -> Logic
+preconditions m@(Model (Opaque (fs, n, nextFd, mp))) cmd = case cmd of
+    Get _ (fd, _) _ ->
+        let (mockHandle, _) = mp Map.! fd
+            ioMode = mockHandleIOMode2 mockHandle
+        -- this condition helps avoiding EBADF errors by trying to read from invalid fds.
+        -- TODO: we could reify this error is FsError and delete this precondition.
+        in Boolean $ ioMode `Prelude.elem` [ReadMode, ReadWriteMode]
+    Put _ (fd, _) _ ->
+        let (mockHandle, _) = mp Map.! fd
+            ioMode = mockHandleIOMode2 mockHandle
+        -- this condition helps avoiding EBADF errors by trying to wrote to invalid fds.
+        -- TODO: we could reify this error is FsError and delete this precondition.
+        in Boolean $ ioMode `Prelude.elem` [WriteMode, ReadWriteMode]
```

Comment:
There's also an `elem` function in the Logic module (https://github.com/advancedtelematic/quickcheck-state-machine/blob/master/src/Test/StateMachine/Logic.hs#L215), which would let you avoid using `Boolean` and give you better counterexamples. It will be renamed to something that doesn't clash with the prelude soon (https://github.com/advancedtelematic/quickcheck-state-machine/issues/241).